openupgrade_scripts/scripts/product/15.0.1.2/upgrade_analysis_work.txt

```diff
@@ -0,0 +1,65 @@
+---Models in module 'product'---
+new model product.label.layout [transient]
+new model report.product.report_producttemplatelabel [abstract]
+new model report.product.report_producttemplatelabel_dymo [abstract]
+# NOTHING TO DO
+
+---Fields in module 'product'---
+product      / product.attribute        / display_type (selection)      : selection_keys is now '['color', 'pills', 'radio', 'select']' ('['color', 'radio', 'select']')
+product      / product.attribute.value  / color (integer)               : NEW hasdefault
+product      / product.packaging        / _order                        : _order is now 'product_id, sequence, id' ('sequence')
+product      / product.product          / _order                        : _order is now 'priority desc, default_code, name, id' ('default_code, name, id')
+# NOTHING TO DO
+
+product      / product.product          / description (text)            : type is now 'html' ('text')
+# DONE: pre-migation: convert text to html
+
+product      / product.product          / product_template_variant_value_ids (many2many): NEW relation: product.template.attribute.value
+product      / product.template         / _order                        : _order is now 'priority desc, name' ('name')
+# NOTHING TO DO
+
+product      / product.template         / description (text)            : type is now 'html' ('text')
+# DONE: pre-migation: convert text to html
+
+product      / product.template         / detailed_type (selection)     : NEW required, selection_keys: ['consu', 'service'], req_default: function, hasdefault
+# DONE: move vale of type to detailed_type
+
+product      / product.template         / type (selection)              : now a function
+# NOTHING TO DO
+
+product      / product.template         / priority (selection)          : NEW selection_keys: ['0', '1'], hasdefault
+# NOTHING TO DO: new fields
+
+product      / product.template.attribute.line / value_count (integer)         : NEW isfunction: function, stored
+# DONE: pre-migration: compute_value_count
+
+product      / product.template.attribute.value / color (integer)               : NEW hasdefault
+# NOTHING TO DO: new field
```

Comment:
Mmm, the default is random: https://github.com/odoo/odoo/blob/15.0/addons/product/models/product_attribute.py#L104-L105. Upon initialization of the field by the ORM, a value is going to be generated once and applied to all attribute lines.  That means that if you're going to do several test migrations, the color is going to be wildly different between two migrations. I feel like that's bound to bug at least someone somewhere. Would you consider setting a neutral value of `0` (grey) explicitely instead?

Reply:
this is a field with random color value, but we can give it a default color, okay i will add this

Reply:
If you feel adventurous, you can assign a *different* random color to each attribute value (I'm sure there is a quick way to do so in SQL). I just want to prevent that *all* attribute values are assigned the same color (which is going to be a different color in each test migration, which may/should raise questions from QA testers)

Reply:
> If you feel adventurous, you can assign a _different_ random color to each attribute value (I'm sure there is a quick way to do so in SQL). I just want to prevent that _all_ attribute values are assigned the same color (which is going to be a different color in each test migration, which may/should raise questions from QA testers)

okay, i get it, i will randomly add color to the records

Reply:
Cool!